### ISSUE TITLE
Deadlock when using QueryResults.stream() from 5+ threads

### ISSUE THREAD

**Issue by tomas-knap** [eclipse/rdf4j#1637]
In our product, there is a method, which runs the code below, so basically it prepares tuple query and then uses QueryResults.stream() to stream over the results of the tuple query evaluation to collect ?s bindings (Note: QueryResults could have been done better in this case, as it anyway always takes only first result, but lets accept it as it is). 

```
...
String queryString = "SELECT ?s " + prepareFromClause(getGraph()) + " WHERE ..."  

TupleQuery tupleQuery = getConnection().prepareTupleQuery(QueryLanguage.SPARQL, queryString);
       
IRI subject = QueryResults.stream(tupleQuery.evaluate()).map(resultSet -> resultSet.getValue("s")).map(value -> (IRI) value).findFirst().orElse(null);

...
```
When I run this code in up to 4 threads in parallel (so I basically call that method at the same moment in up to 4 threads), everything is fine. When I run that in 5 or more threads in parallel, suddenly all threads are WAITING - they are blocked somewhere inside the QueryResults.stream().findFirst(), see the thread state: 

```
engine-starting-execution-4@7038" prio=5 tid=0x38 nid=NA waiting
  java.lang.Thread.State: WAITING
 at sun.misc.Unsafe.park(Unsafe.java:-1)
 at java.util.concurrent.locks.LockSupport.park(LockSupport.java:175)
 at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.await(AbstractQueuedSynchronizer.java:2039)
 at java.util.concurrent.ArrayBlockingQueue.take(ArrayBlockingQueue.java:403)
 at org.eclipse.rdf4j.common.iteration.QueueIteration.getNextElement(QueueIteration.java:139)
 at org.eclipse.rdf4j.common.iteration.LookAheadIteration.lookAhead(LookAheadIteration.java:90)
 at org.eclipse.rdf4j.common.iteration.LookAheadIteration.hasNext(LookAheadIteration.java:52)
 at org.eclipse.rdf4j.common.iteration.IterationWrapper.hasNext(IterationWrapper.java:67)
 at org.eclipse.rdf4j.common.iteration.IterationSpliterator.tryAdvance(IterationSpliterator.java:43)
 at java.util.stream.ReferencePipeline.forEachWithCancel(ReferencePipeline.java:126)
 at java.util.stream.AbstractPipeline.copyIntoWithCancel(AbstractPipeline.java:498)
 at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:485)
 at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:471)
 at java.util.stream.FindOps$FindOp.evaluateSequential(FindOps.java:152)
 at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
 at java.util.stream.ReferencePipeline.findFirst(ReferencePipeline.java:464)
 at cz.cuni.mff.xrg.odcs.commons.app.pipeline.execution.DbExecutionImpl.getLastExecution_aroundBody14(DbExecutionImpl.java:206)
 at cz.cuni.mff.xrg.odcs.commons.app.pipeline.execution.DbExecutionImpl$AjcClosure15.run(DbExecutionImpl.java:1)
 at 
```

And all the other threads, which try to work with the database are blocked in connection.begin() and waiting for the lock. 

When I replace the example from the beginning with the code  below, everything works fine, I can run this method in 50 threads in parallel smoothly:

```
String queryString = "SELECT ?s " + prepareFromClause(getGraph()) + " WHERE ..."      

TupleQuery tupleQuery = getConnection().prepareTupleQuery(QueryLanguage.SPARQL, queryString);

IRI subject = null;
try (TupleQueryResult result = tupleQuery.evaluate()) {
            if (result.hasNext()) {
                Value s = result.next().getValue("s");
                if (s instanceof IRI) {
                    subject = (IRI) s;
                } else {
                   ...
                }
            } 
        }
```
So to summarize: The difference between the working example (the last code example) and the one where we get deadlock (the first code example) is basically the "QueryResults.stream() line" -> if that is replaced with manual iteration, everything works fine. Any ideas?

We use rdf4j 2.3.3, native store over HTTP.

**Comment by hmottestad** [comment]
Does this help?

```
try(.... stream = QueryResults.stream(tupleQuery.evaluate())){
 .... = stream.map(resultSet -> resultSet.getValue("s")).map(value -> (IRI) value).findFirst().orElse(null)
}
```

**Comment by tomas-knap** [comment]
@hmottestad Thanks, indeed, that helps! Btw. Do you have any explanation why such issue occurs so early - I mean after running 5 threads in parallel? Could that be that there is a thread pool for 4 parallel connections only?

**Comment by hmottestad** [comment]
I'm not sure. It would be nice if we could log an error if too many result sets are left open on a connection. Not sure what happens when you close a connection either. 

The whole "everything needs to be closed" is the same for things like JDBC. It's just a bit annoying that Intellij and such haven't figured out a way to warn developers when you forget to close a closeable object. Also a pity that Java streams didn't implement some automatic closing of closeable streams once they've been consumed. 

Are you sharing a connection between threads by any chance?

**Comment by tomas-knap** [comment]
@hmottestad Connection is not shared (although that could be very good explanation of the issue), it is taken from the thread local object, which is different for each thread. So that is something which is quite weird - if those result sets are on different connections, how could they interfere? 

So for me it really looks like there is a maximum number of concurrent requests - if there are more, all threads are parked and are waiting for a slot (remember, the other threads which wanted to also access the DB were actually frozen in connection.begin()). To make it even more complicated, after that "deadlock" happened in our app, it was still possible to access the same database from a second app and also from workbench, so the DB was not completely blocked, just all the requests coming from threads from that one "process" were blocked.

**Comment by hmottestad** [comment]
It would be nice for it not to deadlock. Would also be nice if it logged some error message prior to deadlocking, and maybe even if there was some way of detecting that the query results were no longer in use and should have been closed. Maybe something through finalize or with phantom references.

I am open to sponsorship if this is something your company needs and you want to prioritize a fix. https://github.com/sponsors/hmottestad

**Comment by tomas-knap** [comment]
Thanks @hmottestad, just to double check - there is no hard a priori restriction on maximum number of concurrent requests in rdf4j api/native store, right?

**Comment by tomas-knap** [comment]
@hmottestad Thanks, I think we can close for now
